File: language/real/prolog/allocation_v10.pl
:- use_module(library(readutil)).
:- dynamic lane/6, obligation/3, allocation/3, outcome/3.
:- dynamic packet_id/1, claim_scope/1, budget/1, horizon/1, reserve/1.
:- dynamic normative_prior/1, normative_utility/1, history_signature/1, policy/1.
:- dynamic witness_escape_lane/1.
:- dynamic expected_acrr/1, expected_starvation/1, expected_debt/1.
:- dynamic expected_escape_reallocation/1, expected_common_mode/1.
:- dynamic seen_header/0, seen_end/0.

reset_db :-
    retractall(lane(_,_,_,_,_,_)), retractall(obligation(_,_,_)),
    retractall(allocation(_,_,_)), retractall(outcome(_,_,_)),
    retractall(packet_id(_)), retractall(claim_scope(_)), retractall(budget(_)),
    retractall(horizon(_)), retractall(reserve(_)),
    retractall(normative_prior(_)), retractall(normative_utility(_)),
    retractall(history_signature(_)), retractall(policy(_)),
    retractall(witness_escape_lane(_)),
    retractall(expected_acrr(_)), retractall(expected_starvation(_)),
    retractall(expected_debt(_)), retractall(expected_escape_reallocation(_)),
    retractall(expected_common_mode(_)), retractall(seen_header), retractall(seen_end).

atomize(S,A) :- atom_string(A,S).

parse_line(Line) :-
    normalize_space(string(N), Line),
    ( N="" -> true
    ; sub_string(N,0,1,_,"#") -> true
    ; seen_end -> true
    ; split_string(N," \t"," \t",T),
      ( T=["REALALLOCATE","0.10"] ->
          (parse_tokens(T) -> true ; format(user_error,'PROLOG_PARSE_FAIL tokens=~q line=~s~n',[T,Line]), fail)
      ; seen_header ->
          (parse_tokens(T) -> true ; format(user_error,'PROLOG_PARSE_FAIL tokens=~q line=~s~n',[T,Line]), fail)
      ; format(user_error,'PROLOG_HEADER_REQUIRED line=~s~n',[Line]), fail
      )
    ).

parse_tokens(["REALALLOCATE","0.10"]) :- assertz(seen_header).
parse_tokens(["id",S]) :- atomize(S,A), assertz(packet_id(A)).
parse_tokens(["claim_scope",S]) :- atomize(S,A), assertz(claim_scope(A)).
parse_tokens(["budget",S]) :- number_string(N,S), N>=0, assertz(budget(N)).
parse_tokens(["horizon",S]) :- number_string(N,S), N>=0, assertz(horizon(N)).
parse_tokens(["lane",I,A,C,M,R,K]) :-
    atomize(I,AI), atomize(A,AA), number_string(N,C), N>0,
    atomize(M,AM), member(AM,['MANDATORY','OPTIONAL']),
    atomize(R,AR), member(AR,['REOPENING','EXPLOITATION']),
    atomize(K,AK), member(AK,['GENERATOR','QUERY','REPRESENTATION','INSTRUMENT','RESIDUAL','EXTERNAL_CASE']),
    \+ lane(AI,_,_,_,_,_), assertz(lane(AI,AA,N,AM,AR,AK)).
parse_tokens(["obligation",I,L,D]) :-
    atomize(I,AI), atomize(L,AL), number_string(N,D), N>0,
    \+ obligation(AI,_,_), assertz(obligation(AI,AL,N)).
parse_tokens(["allocate",S,L,U]) :-
    number_string(NS,S), atomize(L,AL), number_string(NU,U), NS>0, NU>=0,
    assertz(allocation(NS,AL,NU)).
parse_tokens(["outcome",S,L,O]) :-
    number_string(NS,S), atomize(L,AL), atomize(O,AO), NS>0,
    member(AO,['NEW','NO_NEW','ESCAPE','NA']), assertz(outcome(NS,AL,AO)).
parse_tokens(["reserve",S]) :- number_string(N,S), N>=0, assertz(reserve(N)).
parse_tokens(["normative_prior",S]) :-
    atomize(S,A), member(A,['NONE','DECLARED']), assertz(normative_prior(A)).
parse_tokens(["normative_utility",S]) :-
    atomize(S,A), member(A,['NONE','DECLARED']), assertz(normative_utility(A)).
parse_tokens(["history_signature",S]) :- atomize(S,A), assertz(history_signature(A)).
parse_tokens(["policy",S]) :- atomize(S,A), assertz(policy(A)).
parse_tokens(["witness_escape_lane",S]) :- atomize(S,A), assertz(witness_escape_lane(A)).
parse_tokens(["authorize_acrr",S]) :- atomize(S,A), assertz(expected_acrr(A)).
parse_tokens(["authorize_starvation",S]) :- atomize(S,A), assertz(expected_starvation(A)).
parse_tokens(["authorize_debt",S]) :- atomize(S,A), assertz(expected_debt(A)).
parse_tokens(["authorize_escape_reallocation",S]) :- atomize(S,A), assertz(expected_escape_reallocation(A)).
parse_tokens(["authorize_common_mode",S]) :- atomize(S,A), assertz(expected_common_mode(A)).
parse_tokens(["END"]) :- assertz(seen_end).

load_packet(File) :-
    reset_db,
    read_file_to_string(File,S,[]),
    split_string(S,"\n","\r",Lines),
    maplist(parse_line,Lines),
    seen_header, seen_end,
    packet_id(_), claim_scope(_), budget(_), horizon(_), reserve(_), lane(_,_,_,_,_,_),
    normative_prior(_), normative_utility(_), history_signature(_), policy(_), witness_escape_lane(W),
    \+ (obligation(_,L,_), \+ lane(L,_,_,'MANDATORY',_,_)),
    \+ (allocation(_,L,_), \+ lane(L,_,_,_,_,_)),
    \+ (allocation(S,_,_), horizon(H), S>H),
    \+ (outcome(_,L,_), \+ lane(L,_,_,_,_,_)),
    \+ (outcome(S,_,_), horizon(H), S>H),
    (W='NONE';lane(W,_,_,_,_,_)).

sum_allocated(T) :-
    findall(U,allocation(_,_,U),Us), sum_list(Us,T).

spend_until(L,Due,T) :-
    findall(U,(allocation(S,L,U),S=<Due),Us), sum_list(Us,T).

role_spend(Role,T) :-
    findall(U,(allocation(_,L,U),lane(L,_,_,_,Role,_)),Us), sum_list(Us,T).

debt_item(I,K) :-
    obligation(I,L,D), horizon(H), D=<H,
    lane(L,_,Cost,'MANDATORY',_,K),
    spend_until(L,D,T), T<Cost.

debt_items(Items,Classes) :-
    findall(X,(debt_item(I,K),atomic_list_concat([I,'@',K],X)),Xs), sort(Xs,Items),
    findall(K,debt_item(_,K),Ks), sort(Ks,Classes).

yes(true,'YES') :- !.
yes(_,'NO').

acrr(true) :-
    reserve(R), R>0, lane(_,_,Cost,_, 'REOPENING',_), Cost=<R, !.
acrr(false).

common_mode(true) :-
    findall(L,lane(L,_,_,_,_,_),Ls0), sort(Ls0,Ls), length(Ls,N), N>1,
    findall(A,lane(_,A,_,_,_,_),As0), sort(As0,As), length(As,M), M<N, !.
common_mode(false).

check_expected(ACRR,Starve,Debt,Escape,Common) :-
    (expected_acrr(X)->X=ACRR;true),
    (expected_starvation(X)->X=Starve;true),
    (expected_debt(X)->atom_number(X,Debt);true),
    (expected_escape_reallocation(X)->X=Escape;true),
    (expected_common_mode(X)->X=Common;true).

join_or_none([], 'NONE').
join_or_none(Xs, A) :- Xs\=[], atomic_list_concat(Xs,'+',A).

run(File) :-
    load_packet(File),
    budget(B), horizon(H), reserve(R),
    sum_allocated(T), Committed is T+R,
    (Committed=<B->BudgetOk='YES';BudgetOk='NO'),
    findall(L,lane(L,_,_,_,_,_),Ls0), sort(Ls0,Ls), length(Ls,LaneCount),
    findall(L,lane(L,_,_,'MANDATORY',_,_),Ms0), sort(Ms0,Ms), length(Ms,MandatoryCount),
    findall(A,lane(_,A,_,_,_,_),As0), sort(As0,As), length(As,AncestryCount),
    common_mode(CMB), yes(CMB,Common),
    (R>0->Nominal='YES';Nominal='NO'),
    acrr(ACRRB), yes(ACRRB,ACRR),
    (ACRR='YES'->ReserveState='LIVE';Nominal='YES'->ReserveState='NOMINAL_ONLY';ReserveState='ABSENT'),
    debt_items(Debts,DebtClasses), length(Debts,DebtCount),
    (DebtCount>0->Starve='YES';Starve='NO'),
    join_or_none(Debts,DebtAtom), join_or_none(DebtClasses,DebtClassAtom),
    role_spend('EXPLOITATION',ExploitSpend), role_spend('REOPENING',ReopenSpend),
    (Starve='YES',ExploitSpend>0,ReopenSpend=:=0->ExplorationStarved='YES';ExplorationStarved='NO'),
    findall(1,outcome(_,_, 'NEW'),NewXs), length(NewXs,NewCount),
    policy(Policy),
    (Policy='GREEDY_YIELD',NewCount>0,ExplorationStarved='YES'->Greedy='YES';Greedy='NO'),
    (outcome(_,_, 'ESCAPE')->Escape='YES';Escape='NO'),
    normative_prior(Prior), normative_utility(Utility),
    (Prior='DECLARED',Utility='DECLARED'->OptScope='DECLARED_MODEL_ONLY';OptScope='NOT_AUTHORIZED'),
    (BudgetOk='YES',Starve='NO',ACRR='YES'->Envelope='YES';Envelope='NO'),
    format('allocation.budget=~w~n',[B]),
    format('allocation.horizon=~w~n',[H]),
    format('allocation.total_allocated=~w~n',[T]),
    format('allocation.reserve_units=~w~n',[R]),
    format('allocation.budget_valid=~w~n',[BudgetOk]),
    format('allocation.lane_count=~w~n',[LaneCount]),
    format('allocation.mandatory_lane_count=~w~n',[MandatoryCount]),
    format('allocation.ancestry_count=~w~n',[AncestryCount]),
    format('allocation.common_mode_detected=~w~n',[Common]),
    format('allocation.nominal_reserve=~w~n',[Nominal]),
    format('allocation.activation_capable_reserve=~w~n',[ACRR]),
    format('allocation.reserve_state=~w~n',[ReserveState]),
    format('allocation.starvation_detected=~w~n',[Starve]),
    format('allocation.exploration_debt_count=~w~n',[DebtCount]),
    format('allocation.exploration_debt_items=~w~n',[DebtAtom]),
    format('allocation.exploration_debt_classes=~w~n',[DebtClassAtom]),
    writeln('allocation.debt_scalar_default=OFF'),
    writeln('allocation.opportunity_cost_mode=VECTOR'),
    format('allocation.exploitation_spend=~w~n',[ExploitSpend]),
    format('allocation.reopening_spend=~w~n',[ReopenSpend]),
    format('allocation.exploration_starved=~w~n',[ExplorationStarved]),
    format('allocation.greedy_yield_starvation_witness=~w~n',[Greedy]),
    format('allocation.escape_detected=~w~n',[Escape]),
    format('allocation.escape_reallocation_required=~w~n',[Escape]),
    writeln('allocation.reopen_on_escape=YES'),
    format('allocation.normative_prior=~w~n',[Prior]),
    format('allocation.normative_utility=~w~n',[Utility]),
    writeln('allocation.hidden_prior_inferred=NO'),
    writeln('allocation.hidden_utility_inferred=NO'),
    writeln('allocation.expected_discovery_value_inferred=NO'),
    writeln('allocation.unique_optimum_authorized=NO'),
    writeln('allocation.universal_next_action=UNIDENTIFIED'),
    format('allocation.local_optimizer_scope=~w~n',[OptScope]),
    writeln('allocation.guidance_mode=SET_VALUED_CONTRACT_RELATIVE'),
    format('allocation.admissible_envelope=~w~n',[Envelope]),
    writeln('allocation.randomization_epistemic_oracle=NO'),
    writeln('allocation.policy_path_provenance=REQUIRED'),
    writeln('allocation.world_optimum_identified=NO'),
    writeln('allocation.world_frontier_complete=NO'),
    writeln('allocation.stopping_rule=FORBIDDEN'),
    check_expected(ACRR,Starve,DebtCount,Escape,Common).

File: language/real/src/allocation_v10.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    env, fs,
    path::Path,
};

#[derive(Clone, Debug)]
struct Lane {
    id: String,
    ancestry: String,
    activation_cost: u64,
    mandatory: bool,
    role: String,
    class: String,
}

#[derive(Clone, Debug)]
struct Obligation {
    id: String,
    lane: String,
    due_step: u64,
}

#[derive(Default, Debug)]
struct Packet {
    id: String,
    claim_scope: String,
    budget: Option<u64>,
    horizon: Option<u64>,
    lanes: BTreeMap<String, Lane>,
    obligations: BTreeMap<String, Obligation>,
    allocations: Vec<(u64, String, u64)>,
    outcomes: Vec<(u64, String, String)>,
    reserve: Option<u64>,
    normative_prior: String,
    normative_utility: String,
    history_signature: String,
    policy: String,
    witness_escape_lane: String,
    expect_acrr: Option<String>,
    expect_starvation: Option<String>,
    expect_debt: Option<String>,
    expect_escape_reallocation: Option<String>,
    expect_common_mode: Option<String>,
}

fn toks(s: &str) -> Vec<String> {
    s.split_whitespace().map(|x| x.trim_matches('"').to_string()).collect()
}

fn parse_u64(x: &str, what: &str) -> Result<u64, String> {
    x.parse::<u64>().map_err(|_| format!("invalid {what}: {x}"))
}

fn parse(path: &Path) -> Result<Packet, String> {
    let text = fs::read_to_string(path).map_err(|e| e.to_string())?;
    let mut p = Packet::default();
    let mut started = false;
    let mut ended = false;

    for (i, raw) in text.lines().enumerate() {
        let line = raw.trim();
        if line.is_empty() || line.starts_with('#') { continue; }
        let t = toks(line);
        if t.is_empty() { continue; }
        if t[0] == "REALALLOCATE" {
            if t.len() != 2 || t[1] != "0.10" { return Err("expected REALALLOCATE 0.10".into()); }
            started = true;
            continue;
        }
        if !started { return Err("packet must begin REALALLOCATE 0.10".into()); }
        if t[0] == "END" { ended = true; break; }

        match t[0].as_str() {
            "id" if t.len() == 2 => p.id = t[1].clone(),
            "claim_scope" if t.len() == 2 => p.claim_scope = t[1].clone(),
            "budget" if t.len() == 2 => p.budget = Some(parse_u64(&t[1], "budget")?),
            "horizon" if t.len() == 2 => p.horizon = Some(parse_u64(&t[1], "horizon")?),
            "lane" if t.len() == 7 => {
                let activation_cost = parse_u64(&t[3], "activation cost")?;
                if activation_cost == 0 { return Err("activation cost must be positive".into()); }
                let obligation = t[4].to_uppercase();
                if !["MANDATORY", "OPTIONAL"].contains(&obligation.as_str()) {
                    return Err(format!("invalid lane obligation: {}", t[4]));
                }
                let role = t[5].to_uppercase();
                if !["REOPENING", "EXPLOITATION"].contains(&role.as_str()) {
                    return Err(format!("invalid lane role: {}", t[5]));
                }
                let class = t[6].to_uppercase();
                if !["GENERATOR", "QUERY", "REPRESENTATION", "INSTRUMENT", "RESIDUAL", "EXTERNAL_CASE"].contains(&class.as_str()) {
                    return Err(format!("invalid lane class: {}", t[6]));
                }
                if p.lanes.contains_key(&t[1]) { return Err(format!("duplicate lane: {}", t[1])); }
                p.lanes.insert(t[1].clone(), Lane {
                    id: t[1].clone(),
                    ancestry: t[2].clone(),
                    activation_cost,
                    mandatory: obligation == "MANDATORY",
                    role,
                    class,
                });
            }
            "obligation" if t.len() == 4 => {
                let due_step = parse_u64(&t[3], "obligation due step")?;
                if p.obligations.contains_key(&t[1]) { return Err(format!("duplicate obligation: {}", t[1])); }
                p.obligations.insert(t[1].clone(), Obligation {
                    id: t[1].clone(),
                    lane: t[2].clone(),
                    due_step,
                });
            }
            "allocate" if t.len() == 4 => p.allocations.push((
                parse_u64(&t[1], "allocation step")?,
                t[2].clone(),
                parse_u64(&t[3], "allocation units")?,
            )),
            "outcome" if t.len() == 4 => {
                let state = t[3].to_uppercase();
                if !["NEW", "NO_NEW", "ESCAPE", "NA"].contains(&state.as_str()) {
                    return Err(format!("invalid outcome state: {}", t[3]));
                }
                p.outcomes.push((parse_u64(&t[1], "outcome step")?, t[2].clone(), state));
            }
            "reserve" if t.len() == 2 => p.reserve = Some(parse_u64(&t[1], "reserve")?),
            "normative_prior" if t.len() == 2 => {
                let x = t[1].to_uppercase();
                if !["NONE", "DECLARED"].contains(&x.as_str()) { return Err(format!("invalid normative_prior: {}", t[1])); }
                p.normative_prior = x;
            }
            "normative_utility" if t.len() == 2 => {
                let x = t[1].to_uppercase();
                if !["NONE", "DECLARED"].contains(&x.as_str()) { return Err(format!("invalid normative_utility: {}", t[1])); }
                p.normative_utility = x;
            }
            "history_signature" if t.len() == 2 => p.history_signature = t[1].clone(),
            "policy" if t.len() == 2 => p.policy = t[1].to_uppercase(),
            "witness_escape_lane" if t.len() == 2 => p.witness_escape_lane = t[1].clone(),
            "authorize_acrr" if t.len() == 2 => p.expect_acrr = Some(t[1].to_uppercase()),
            "authorize_starvation" if t.len() == 2 => p.expect_starvation = Some(t[1].to_uppercase()),
            "authorize_debt" if t.len() == 2 => p.expect_debt = Some(t[1].clone()),
            "authorize_escape_reallocation" if t.len() == 2 => p.expect_escape_reallocation = Some(t[1].to_uppercase()),
            "authorize_common_mode" if t.len() == 2 => p.expect_common_mode = Some(t[1].to_uppercase()),
            _ => return Err(format!("{}:{} malformed command: {line}", path.display(), i + 1)),
        }
    }

    if !ended { return Err("missing END".into()); }
    if p.id.is_empty() || p.claim_scope.is_empty() || p.budget.is_none() || p.horizon.is_none()
        || p.reserve.is_none() || p.lanes.is_empty() || p.normative_prior.is_empty()
        || p.normative_utility.is_empty() || p.history_signature.is_empty()
        || p.policy.is_empty() || p.witness_escape_lane.is_empty()
    {
        return Err("missing required field".into());
    }

    let horizon = p.horizon.unwrap();
    for o in p.obligations.values() {
        let Some(lane) = p.lanes.get(&o.lane) else {
            return Err(format!("unknown lane for obligation {}: {}", o.id, o.lane));
        };
        if !lane.mandatory { return Err(format!("obligation {} targets non-mandatory lane {}", o.id, o.lane)); }
        if o.due_step == 0 { return Err(format!("obligation {} has zero due step", o.id)); }
    }
    for (step, lane, _) in &p.allocations {
        if *step == 0 || *step > horizon { return Err(format!("allocation step out of horizon: {step}")); }
        if !p.lanes.contains_key(lane) { return Err(format!("unknown lane in allocation: {lane}")); }
    }
    for (step, lane, _) in &p.outcomes {
        if *step == 0 || *step > horizon { return Err(format!("outcome step out of horizon: {step}")); }
        if !p.lanes.contains_key(lane) { return Err(format!("unknown lane in outcome: {lane}")); }
    }
    if p.witness_escape_lane != "NONE" && !p.lanes.contains_key(&p.witness_escape_lane) {
        return Err(format!("unknown witness escape lane: {}", p.witness_escape_lane));
    }
    Ok(p)
}

fn yes(b: bool) -> &'static str { if b { "YES" } else { "NO" } }

fn spend_until(p: &Packet, lane: &str, due: u64) -> u64 {
    p.allocations.iter()
        .filter(|(step, l, _)| *step <= due && l == lane)
        .fold(0u64, |acc, (_, _, u)| acc.saturating_add(*u))
}

fn spend_role(p: &Packet, role: &str) -> u64 {
    p.allocations.iter()
        .filter(|(_, lane, _)| p.lanes.get(lane).map(|x| x.role.as_str()) == Some(role))
        .fold(0u64, |acc, (_, _, u)| acc.saturating_add(*u))
}

fn analyze(p: &Packet) -> BTreeMap<String, String> {
    let mut a = BTreeMap::new();
    let budget = p.budget.unwrap();
    let horizon = p.horizon.unwrap();
    let reserve = p.reserve.unwrap();
    let total_allocated = p.allocations.iter().fold(0u64, |acc, (_, _, u)| acc.saturating_add(*u));
    let budget_valid = total_allocated.saturating_add(reserve) <= budget;

    let lane_count = p.lanes.len();
    let mandatory_count = p.lanes.values().filter(|l| l.mandatory).count();
    let ancestries: BTreeSet<_> = p.lanes.values().map(|l| l.ancestry.clone()).collect();
    let common_mode = lane_count > 1 && ancestries.len() < lane_count;

    let reopening_lanes: Vec<_> = p.lanes.values().filter(|l| l.role == "REOPENING").collect();
    let nominal_reserve = reserve > 0;
    let acrr = nominal_reserve && reopening_lanes.iter().any(|l| l.activation_cost <= reserve);
    let reserve_state = if acrr { "LIVE" } else if nominal_reserve { "NOMINAL_ONLY" } else { "ABSENT" };

    let mut debts = Vec::new();
    let mut debt_classes = BTreeSet::new();
    for o in p.obligations.values() {
        if o.due_step > horizon { continue; }
        let lane = &p.lanes[&o.lane];
        if spend_until(p, &o.lane, o.due_step) < lane.activation_cost {
            debts.push(format!("{}@{}", o.id, lane.class));
            debt_classes.insert(lane.class.clone());
        }
    }
    debts.sort();
    let debt_count = debts.len();
    let starvation = debt_count > 0;

    let escape = p.outcomes.iter().any(|(_, _, s)| s == "ESCAPE");
    let new_count = p.outcomes.iter().filter(|(_, _, s)| s == "NEW").count();
    let exploitation_spend = spend_role(p, "EXPLOITATION");
    let reopening_spend = spend_role(p, "REOPENING");
    let exploration_starved = starvation && exploitation_spend > 0 && reopening_spend == 0;
    let greedy_yield_starvation = p.policy == "GREEDY_YIELD" && new_count > 0 && exploration_starved;
    let admissible = budget_valid && !starvation && acrr;

    a.insert("allocation.budget".into(), budget.to_string());
    a.insert("allocation.horizon".into(), horizon.to_string());
    a.insert("allocation.total_allocated".into(), total_allocated.to_string());
    a.insert("allocation.reserve_units".into(), reserve.to_string());
    a.insert("allocation.budget_valid".into(), yes(budget_valid).into());
    a.insert("allocation.lane_count".into(), lane_count.to_string());
    a.insert("allocation.mandatory_lane_count".into(), mandatory_count.to_string());
    a.insert("allocation.ancestry_count".into(), ancestries.len().to_string());
    a.insert("allocation.common_mode_detected".into(), yes(common_mode).into());
    a.insert("allocation.nominal_reserve".into(), yes(nominal_reserve).into());
    a.insert("allocation.activation_capable_reserve".into(), yes(acrr).into());
    a.insert("allocation.reserve_state".into(), reserve_state.into());
    a.insert("allocation.starvation_detected".into(), yes(starvation).into());
    a.insert("allocation.exploration_debt_count".into(), debt_count.to_string());
    a.insert("allocation.exploration_debt_items".into(), if debts.is_empty() { "NONE".into() } else { debts.join("+") });
    a.insert("allocation.exploration_debt_classes".into(), if debt_classes.is_empty() { "NONE".into() } else { debt_classes.into_iter().collect::<Vec<_>>().join("+") });
    a.insert("allocation.debt_scalar_default".into(), "OFF".into());
    a.insert("allocation.opportunity_cost_mode".into(), "VECTOR".into());
    a.insert("allocation.exploitation_spend".into(), exploitation_spend.to_string());
    a.insert("allocation.reopening_spend".into(), reopening_spend.to_string());
    a.insert("allocation.exploration_starved".into(), yes(exploration_starved).into());
    a.insert("allocation.greedy_yield_starvation_witness".into(), yes(greedy_yield_starvation).into());
    a.insert("allocation.escape_detected".into(), yes(escape).into());
    a.insert("allocation.escape_reallocation_required".into(), yes(escape).into());
    a.insert("allocation.reopen_on_escape".into(), "YES".into());
    a.insert("allocation.normative_prior".into(), p.normative_prior.clone());
    a.insert("allocation.normative_utility".into(), p.normative_utility.clone());
    a.insert("allocation.hidden_prior_inferred".into(), "NO".into());
    a.insert("allocation.hidden_utility_inferred".into(), "NO".into());
    a.insert("allocation.expected_discovery_value_inferred".into(), "NO".into());
    a.insert("allocation.unique_optimum_authorized".into(), "NO".into());
    a.insert("allocation.universal_next_action".into(), "UNIDENTIFIED".into());
    a.insert("allocation.local_optimizer_scope".into(),
        if p.normative_prior == "DECLARED" && p.normative_utility == "DECLARED" { "DECLARED_MODEL_ONLY".into() } else { "NOT_AUTHORIZED".into() });
    a.insert("allocation.guidance_mode".into(), "SET_VALUED_CONTRACT_RELATIVE".into());
    a.insert("allocation.admissible_envelope".into(), yes(admissible).into());
    a.insert("allocation.randomization_epistemic_oracle".into(), "NO".into());
    a.insert("allocation.policy_path_provenance".into(), "REQUIRED".into());
    a.insert("allocation.world_optimum_identified".into(), "NO".into());
    a.insert("allocation.world_frontier_complete".into(), "NO".into());
    a.insert("allocation.stopping_rule".into(), "FORBIDDEN".into());
    a
}

fn validate(p: &Packet, a: &BTreeMap<String, String>) -> Result<(), String> {
    let checks = [
        (&p.expect_acrr, "allocation.activation_capable_reserve"),
        (&p.expect_starvation, "allocation.starvation_detected"),
        (&p.expect_debt, "allocation.exploration_debt_count"),
        (&p.expect_escape_reallocation, "allocation.escape_reallocation_required"),
        (&p.expect_common_mode, "allocation.common_mode_detected"),
    ];
    for (expected, key) in checks {
        if let Some(x) = expected {
            if a[key] != *x { return Err(format!("{key} mismatch expected {x} got {}", a[key])); }
        }
    }
    Ok(())
}

fn emit(p: &Packet, a: &BTreeMap<String, String>) -> String {
    let mut s = format!(
        "REAL-ALLOCATION=0.10\nid={}\nclaim_scope={}\nhistory_signature={}\npolicy={}\n",
        p.id, p.claim_scope, p.history_signature, p.policy
    );
    for key in [
        "allocation.budget","allocation.horizon","allocation.total_allocated","allocation.reserve_units",
        "allocation.budget_valid","allocation.lane_count","allocation.mandatory_lane_count","allocation.ancestry_count",
        "allocation.common_mode_detected","allocation.nominal_reserve","allocation.activation_capable_reserve",
        "allocation.reserve_state","allocation.starvation_detected","allocation.exploration_debt_count",
        "allocation.exploration_debt_items","allocation.exploration_debt_classes","allocation.debt_scalar_default",
        "allocation.opportunity_cost_mode","allocation.exploitation_spend","allocation.reopening_spend",
        "allocation.exploration_starved","allocation.greedy_yield_starvation_witness","allocation.escape_detected",
        "allocation.escape_reallocation_required","allocation.reopen_on_escape","allocation.normative_prior",
        "allocation.normative_utility","allocation.hidden_prior_inferred","allocation.hidden_utility_inferred",
        "allocation.expected_discovery_value_inferred","allocation.unique_optimum_authorized",
        "allocation.universal_next_action","allocation.local_optimizer_scope","allocation.guidance_mode",
        "allocation.admissible_envelope","allocation.randomization_epistemic_oracle",
        "allocation.policy_path_provenance","allocation.world_optimum_identified",
        "allocation.world_frontier_complete","allocation.stopping_rule",
    ] {
        s.push_str(&format!("{key}={}\n", a[key]));
    }
    s.push_str(&format!("witness.escape_lane={}\n", p.witness_escape_lane));
    s.push_str("allocation.meaning=CONTRACT_RELATIVE_OPEN_FRONTIER_ATTENTION_GOVERNANCE\n");
    s.push_str("admissible_does_not_mean_optimal=true\n");
    s.push_str("procedural_coverage_does_not_mean_frontier_completeness=true\n");
    s.push_str("finite_budget_does_not_price_unknown_rivals=true\n");
    s.push_str("final_truth_distance=UNIDENTIFIED\n");
    s
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    if args.is_empty() {
        eprintln!("usage: real-v10-allocate <packet.real> [...]");
        std::process::exit(2);
    }
    for f in args {
        match parse(Path::new(&f)) {
            Ok(p) => {
                let a = analyze(&p);
                if let Err(e) = validate(&p, &a) {
                    eprintln!("{e}");
                    std::process::exit(1);
                }
                print!("{}", emit(&p, &a));
            }
            Err(e) => {
                eprintln!("{e}");
                std::process::exit(1);
            }
        }
    }
}
